src/completion_items_cache.ts

```diff
@@ -15,20 +17,65 @@ export function refresh() {
     workspaceFolders.forEach(workspaceFolder => {
-        const relativePattern = new vscode.RelativePattern(workspaceFolder, "**/*.ts");
-        vscode.workspace.findFiles(relativePattern).then(
-            uris => {
-                const completionItems = uris.map(uri => uriToCompletionItem(uri));
-                workspaceCompletionItems[workspaceFolder.index] = new vscode.CompletionList(completionItems, false);
-            },
-            error => {
-                console.error("Error in updateWorkspaceCompletionItems: " + error);
+        const tsconfigPattern = new vscode.RelativePattern(workspaceFolder, "**/tsconfig.json");
+        const typescriptPattern = new vscode.RelativePattern(workspaceFolder, "**/*.ts");
+        vscode.workspace
+            .findFiles(tsconfigPattern)
+            .then(_tsconfigUrisToBaseUrlMap(workspaceFolder), error => {
+                console.error(`Error while working with **/tsconfig.json files ${error}`);
+                return {};
+            })
+            .then(baseUrlMap => {
+                vscode.workspace.findFiles(typescriptPattern).then(
+                    uris => {
+                        const completionItems = uris.map(uri => uriToCompletionItem(uri, baseUrlMap));
+                        workspaceCompletionItems[workspaceFolder.index] = new vscode.CompletionList(
+                            completionItems,
+                            false
+                        );
+                    },
+                    error => {
+                        console.error(`Error in updateWorkspaceCompletionItems: ${error}`);
+                    }
+                );
             });
     });
 }
 
+const _tsconfigDocumentToBaseUrl = (tsconfigDoc: vscode.TextDocument): string | null => {
+    const tsconfigObj = JSON.parse(stripJsonComments(tsconfigDoc.getText()));
+    if ("compilerOptions" in tsconfigObj && "baseUrl" in tsconfigObj["compilerOptions"]) {
+        return <string>tsconfigObj["compilerOptions"]["baseUrl"];
+    }
+    return null;
+};
+
+const _tsconfigUrisToBaseUrlMap = (workspaceFolder: vscode.WorkspaceFolder) => (
+    uris: vscode.Uri[]
+): Thenable<Record<string, string>> => {
+    const recordPromises = Promise.all(
+        uris.map(tsconfigUri =>
+            vscode.workspace
+                .openTextDocument(tsconfigUri)
+                .then(_tsconfigDocumentToBaseUrl)
+                .then(maybeBaseUrl =>
+                    maybeBaseUrl
+                        ? {
+                              [Path.relative(workspaceFolder.uri.path, Path.dirname(tsconfigUri.path))]: maybeBaseUrl
+                          }
+                        : null
+                )
+        )
+    );
+    return recordPromises.then(
+        records =>
+            // We filter out nulls so it's pretty safe to cast.
+            <Record<string, string>>records.filter(r => r !== null).reduce((acc, r) => ({ ...acc, ...r }))
```

Comment:
can you recommend some transformer function which will yield us with one part of union type (excluding `null` for this case) so that we don't cast?

Reply:
The way I can see to do this would be something like this where you just make the record and then fill it in with forEach. I don't think filter can restrict the type from my playing with it.  http://www.typescriptlang.org/play/index.html#src=type%20Data%20%3D%20%7B%0A%20%20%20%20pathFromWorkspaceRoot%3A%20string%3B%0A%20%20%20%20baseUrl%3A%20string%3B%0A%7D%3B%0A%0Adeclare%20var%20calculateData%3A%20(tsconfigUrl%3A%20string)%20%3D%3E%20Promise%3CData%20%7C%20void%3E%3B%0Adeclare%20var%20tsconfigUrls%3A%20string%5B%5D%3B%0Adeclare%20var%20isSome%3A%20%3CT%3E(t%3A%20T%20%7C%20void)%20%3D%3E%20t%20is%20T%3B%20%0A%0APromise.all(tsconfigUrls.map(url%20%3D%3E%20calculateData(url))).then(values%20%3D%3E%20%7B%0A%20%20%20%20let%20record%3A%20Record%3Cstring%2C%20string%3E%20%3D%20%7B%7D%3B%0A%20%20%20%20values.forEach(value%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20if%20(isSome(value))%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20record%5Bvalue.pathFromWorkspaceRoot%5D%20%3D%20value.baseUrl%3B%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D)%0A%20%20%20%20return%20record%3B%0A%7D)%20

Reply:
I see... Considering this, I'd stay with a current approach to avoid boilerplate of imperative declare-mutate-return approach